A69-crl-enhancements.md

```diff
@@ -0,0 +1,226 @@
+1. The CRL Providers simply provide CRLs. The actual verification will be done by other pieces of the code (these are already written).
+1. We cannot know users' PKI details, so our implementations should not enforce non-X509 metadata mattering (for example, file names). A user could certainly implement their own provider that cares about filenames.
+1. We will not support OSCP-style checking (OSCP is an alternative to CRLs, not another form of CRLs).
+
+![Basic Flow Diagram](A69_graphics/basic_diagram.png)
+
+CrlProvider Interface Pseudocode:
+```
+public:
+// Get the CRL that is associated with a cert. Provides good semantic
+// consistency as most X509 APIs that try to get a CRL expect to get a cert
+// passed in. In each language we may use an abstraction here that makes it easy
+// to get CRL-lookup information from the certificate.
+// Return value expected to be as read-only 
+// Returns null if there isn't an associated CRL (as null is different than an
+// empty CRL)
+  CRL Crl(certficate/certificateInfo)
+private:
+  // Implementation for getting and storing CRLs
+```
+
+Example feature use:
+```
+opts = TlsOptions()
+provider = CrlDirectoryReloaderProvider("/path/to/crl/directory", Duration(60 seconds))
+opts.provider = provider
+server = Server(opts)
+```
+
+### Map Content and API Outcomes
+
+The diagram below describes the behavior of the API and why this or that outcome is produced. Here, "Map" is a generic way to represent some in-memory representation of the CRLs in a key-value store. “CRL Provider API Return” refers to the CRL return of the `CrlProvider` in the Pseudocode above.
+![Basic API Outcomes](A69_graphics/basic_table.png)
+
+The pseudo code snippet below illustrates the same idea:
+```
+<pseudo code in the handshaker when verifying cert>
+
+RevocationStatus GetRevocationStatus(cert, crlProvider, settings) {
+  crl = crlProvider.Crl(cert)
+  revocationStatus = checkRevocation(crl, cert) // RevocationUndetermined for null CRL
+  if revocationStatus is RevocationUndetermined {
+      return settings.FailOpen ? Unrevoked : Revoked
+  }
+  return revocationStatus 
+}
+```
+
+### Provider Implementation - Static Provider
+
+This is a simple provider that takes in all CRLs to be used as raw strings during initialization, then returns them in the appropriate format when called. The CRLs will be stored in a mapping of a hash of the issuer to the CRL content.
+
+### Provider Implementation - Directory Provider
+This provider will periodically read CRL files in a given directory and update gRPCs internal representation of those CRLs. We expect this will be heavily used, as a directory of CRLs is very common for X509 CRL files. The CRLs will be stored in a mapping of a hash of the issuer to the CRL content - `map<issuer_hash, crl_object>`. The initial read of the directory and filling of this map will be synchronous during server startup when the provider is created.
```

Comment:
It's probably worth explicitly stating that the reason that the initial read is synchronous is to ensure that the CRL data is already loaded before the server starts up, so that we don't delay handshakes while waiting for the CRLs to load.  We should explicitly state that the creation does *not* fail if there is an error in this initial load.